### ISSUE TITLE
uwsgi adds Connection: close  --> client is rated as a bot

### ISSUE THREAD

**Issue by return42** [searxng/searxng#2892]
Related to an old filtron issue, `Connection: close` is added to the HTTP headers .. [The http_connection method evaluates a request as the request of a bot if the header is set to close](https://docs.searxng.org/admin/searx.botdetection.html#method-http-connection). I can't understand how this setting could survive here for so long ... since we replaced filtron with the limiter, clients are classified as bot if they give a `Connection: close`.

https://github.com/searxng/searxng/blob/e509cb7c459a59c6793922bce6ce0cc57006aac6/dockerfiles/uwsgi.ini#L37-L39

Pointed out by @samsapti on [Matrix](https://matrix.to/#/!uotInzBKwlMfhtqZvW:matrix.org/$rvmUvX6jbfzz05XMcruNQjq1jepz1Pzm2vbv7MeMvec?via=matrix.org&via=envs.net&via=frei.chat)

Related:

- https://github.com/searx/searx-docker/issues/24

**Comment by catfluoride** [comment]
I guess this is only true for docker installs?

**Comment by dalf** [comment]
The [Connection header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Connection) works with the HTTP/1 protocol, it does not work with HTTP/2 and HTTP/3. According to MDN, the 

> * `close` indicates that either the client or the server would like to close the connection. This is the default on HTTP/1.0 requests.
> * any comma-separated list of HTTP headers [Usually `keep-alive` only]. Indicates that the client would like to keep the connection open. Keeping a connection open is the default on HTTP/1.1 requests.

So the check `Connection=closed` means if the HTTP use the HTTP/1.0 protocol and does not use HTTP/1.1 (keep alive), then it is a bot.

Now a workflow:

```mermaid
sequenceDiagram
    Browser->>Nginx: (a) most probably HTTP/2 connection 
    Nginx->>Uwsgi: (b) HTTP/1.0 connection, Nginx sets "Connection:close"
    Uwsgi->>SearXNG: (c) WSGI request
    SearXNG-->>Uwsgi: (d) WSGI response
    Uwsgi-->>Nginx: (e) uwsgi.ini : "add-header = Connection: close"
    Nginx-->>Browser: (f) HTTP/2 response
```

b:
* [Nginx sets the header `Connection` to `close` ](https://docs.nginx.com/nginx/admin-guide/web-server/reverse-proxy/#passing-request-headers) according to its documentation. I don't know for Apache and Caddy.

d:
* WSGI response tells it only support HTTP/1.0 not HTTP/1.1. See [settings.yml](https://github.com/searxng/searxng/blob/e509cb7c459a59c6793922bce6ce0cc57006aac6/searx/settings.yml#L86) and [flaskfix.py](https://github.com/searxng/searxng/blob/e509cb7c459a59c6793922bce6ce0cc57006aac6/searx/flaskfix.py#L76). [Related commit](https://github.com/searx/searx/commit/a98bbefbcf89ad8d5959c0ddd1c02ab65b92913d)

e:
* `uwsgi.ini` contains "add-header = Connection: close"

---

I feel [this check](https://github.com/searxng/searxng/blob/master/searx/botdetection/http_connection.py) is useless at least for Nginx without additional configuration, but I remember some text browser were blocked by this rule, so perhaps Caddy forwards the `Connection` header (?).

For Nginx it can work if the `Connection` is forwarded from (a) to (b) with something like this:
`proxy_set_header X-Forwarded-Connection $upstream_http_connection`

It requires to update all existing Nginx/Apache/Caddy configurations.

---

TLDR: 
* I think [this check](https://github.com/searxng/searxng/blob/master/searx/botdetection/http_connection.py) should be removed.
* I don't know for `add-header = Connection: close` but it won't solve the botdetection issue with the `Connection` header.

**Comment by return42** [comment]
Thanks for the deep analysis ..

> I think [this check](https://github.com/searxng/searxng/blob/master/searx/botdetection/http_connection.py) should be removed.

... done!

**Comment by dalf** [comment]
> Thanks for the deep analysis ..

I've tried to recover the various info I could remember where to find, but I'm nearly sure I forget some details.
